pydis_site/apps/home/signals.py

```diff
@@ -0,0 +1,159 @@
+    for by choosing a Sender, if required.
+
+    Signals themselves define a set of arguments that they will provide to Receivers when the
+    Signal is sent. They are always keyword arguments, and Django recommends that all Receiver
+    functions accept them as `**kwargs` (and will supposedly error if you don't do this),
+    supposedly because Signals can change in the future and your receivers should still work.
+
+    Signals do provide a list of their arguments when they're initially constructed, but this
+    is purely for documentation purposes only and Django does not enforce it.
+
+    The Django Signals docs are here: https://docs.djangoproject.com/en/2.2/topics/signals/
+    """
+
+    def __init__(self):
+        post_save.connect(self.model_updated, sender=DiscordUser)
+
+        pre_social_login.connect(self.social_account_updated)
+        social_account_added.connect(self.social_account_updated)
+        social_account_updated.connect(self.social_account_updated)
+        social_account_removed.connect(self.social_account_removed)
+
+        user_logged_in.connect(self.user_logged_in)
+
+    def user_logged_in(self, sender: Type[DjangoUser], **kwargs) -> None:
+        """Handles signals relating to Allauth logins."""
+        user: DjangoUser = kwargs["user"]
+
+        try:
+            account: SocialAccount = SocialAccount.objects.get(
+                user=user, provider=DiscordProvider.id
+            )
+        except SocialAccount.DoesNotExist:
+            return  # User's never linked a Discord account
+
+        try:
+            discord_user: DiscordUser = DiscordUser.objects.get(id=int(account.uid))
+        except DiscordUser.DoesNotExist:
+            return
+
+        self._apply_groups(discord_user, account)
+
+    def social_account_updated(self, sender: Type[SocialLogin], **kwargs) -> None:
+        """Handle signals relating to new/existing social accounts."""
+        social_login: SocialLogin = kwargs["sociallogin"]
+
+        account: SocialAccount = social_login.account
+        provider: Provider = account.get_provider()
+
+        if not isinstance(provider, DiscordProvider):
+            return
+
+        try:
+            user: DiscordUser = DiscordUser.objects.get(id=int(account.uid))
+        except DiscordUser.DoesNotExist:
+            return
+
+        self._apply_groups(user, account)
+
+    def social_account_removed(self, sender: Type[SocialLogin], **kwargs) -> None:
+        """Handle signals relating to removal of social accounts."""
+        account: SocialAccount = kwargs["socialaccount"]
+        provider: Provider = account.get_provider()
+
+        if not isinstance(provider, DiscordProvider):
+            return
+
+        try:
+            user: DiscordUser = DiscordUser.objects.get(id=int(account.uid))
+        except DiscordUser.DoesNotExist:
+            return
+
+        self._apply_groups(user, account, True)
```

Comment:
The kwarg is a bit hard to figure out without reading the function, would be nice to

```suggestion
        self._apply_groups(user, account, deletion=True)
```